### DISCUSSION TITLE
Single Request, multiple sessions - Best practices concerning Load Document Speed?

### DISCUSSION THREAD

**Discussion by marcopasveer** [ravendb/ravendb#13193] [Q&A] [answered]
We are using an Asp.Net core web application with RavenDB as our database running on 1 core, 6GB memory VM. 
When our users make a request for a product we sometimes see `LoadAsync(documentId)` times of over a second for simple documents (less than 1 KB).
I was wonderering what the best practices are concerning the following situation: (how we can speed this up)

User makes a request: 
1 - We check a slug table with product identifiers to see which product the user whishes to see.
2 - Fetch the product using the identifier from the slug dto.
3 - Fetch the shoppingcart using a cookie from the user.
4 - Fetch related products, images, and other related data.
5 - Each fetched document is used in a separate view component or partial view.

Basically for each request we open up to 10 separate new sessions, **using the same documentstore**, and Load (Not Query) single documents.
Fetched data is allowed to be stale. Quick and dirty read is only needed.

- Should we inject a single session or request executor via DI for the entire Request?
- Should we use LoadBalanceBehavior.UseSessionContext?
- Would it be better/faster to use the REST api? (This feels faster, not having to create the initial session)
- Would it be better/faster to create an index, and use the query `FirstOrDefault(a => a.DocumentId == documentId)`

Due to the structure of the application it would be difficult to create an index where we combine the related documents together.

Simplified Code: 
```cs

//***** In StartUp.cs:
services.AddSingleton(new DocumentStore
{
    Urls = "Two different nodes",
    Database = "Single database",
    Conventions = new DocumentConventions
    {
        ReadBalanceBehavior = ReadBalanceBehavior.RoundRobin,
    }
});

// The next part is simplified, its actually split up in various methods.

Slug slug;
Product product;
ShoppingCart shoppingcart; 
IDocumentStore _store; // injected by DI via services.AddSingleton(new DocumentStore) in the StartUp.cs;

//***** Fetch the slug dto the user requested. The product id matching the slug is returned.
using (var session = _store.OpenAsyncSession(new SessionOptions { NoTracking = true }))
    slug = await session.LoadAsync<Slug>($"slug{slug}");

//***** Use the product id from the slug dto to fetch the product.
using (var session = _store.OpenAsyncSession(new SessionOptions { NoTracking = true }))
    product = await session.LoadAsync<Product>($"product:{productId}");

//***** Fetch the shoppingcart. (Id stored in a cookie)
using (var session = _store.OpenAsyncSession(new SessionOptions { NoTracking = true }))
    shoppingcart = await session.LoadAsync<ShoppingCart>($"shoppingcart:{shoppingcartId}");

// 7 more _store.OpenAsyncSessions(); for other querying related data.

```

**Comment by ayende** [comment, marked as answer]
Use a single document session per request. There is no need to drop to the REST layer.
However, you should also take advantage of both includes and multi load, that will give you a single remote call to load all the documents.

**Comment by marcopasveer** [comment]
Thanks, 

If other people are having similar issues, this blog post might also help: https://ayende.com/blog/187906-B/using-ravendb-unit-of-work-and-net-core-mvc
